### PR TITLE
[#1385] Adopt `AccountBalance` fields in Kotlin layer

### LINKED ISSUE #1385: Adopt `AccountBalance` fields in Kotlin layer
## Is your feature request related to a problem? Please describe.
This covers the necessary Kotlin side changes for https://github.com/Electric-Coin-Company/zcash-android-wallet-sdk/pull/1380.

### PR DESCRIPTION
**Pull request by HonzaR**
- Adds the Kotlin side changes for #1380
- Changed WalletBalance to contain these three fields: available, changePending, and valuePending, instead of total and available, and change the transparent flow to StateFlow<Zatoshi>, as we don't distinguish there.
- Related connected APIs changed
- Closes #1385

> **Note**
>This code review checklist is intended to serve as a starting point for the author and reviewer, although it may not be appropriate for all types of changes (e.g. fixing a spelling typo in documentation).  For more in-depth discussion of how we think about code review, please see [Code Review Guidelines](../blob/main/docs/CODE_REVIEW_GUIDELINES.md).

# Author

- [x] **Self-review** your own code in GitHub's web interface[^1]
- [ ] Add **automated tests** as appropriate
- [ ] Update the [**manual tests**](../blob/main/docs/testing/manual_testing)[^2] as appropriate
- [ ] Check the **code coverage**[^3] report for the automated tests
- [x] Update **documentation** as appropriate (e.g [README.md](../blob/main/README.md), [Architecture.md](../blob/main/docs/Architecture.md), etc.)
- [x] **Run the demo app** and try the changes
- [x] Pull in the latest changes from the **main** branch and **squash** your commits before assigning a reviewer[^4]

# Reviewer

- [ ] Check the code with the [Code Review Guidelines](../blob/main/docs/CODE_REVIEW_GUIDELINES.md) **checklist**
- [ ] Perform an **ad hoc review**[^5]
- [ ] Review the **automated tests**
- [ ] Review the **manual tests**
- [ ] Review the **documentation**, [README.md](../blob/main/README.md), [Architecture.md](/blob/main/docs/Architecture.md), etc. as appropriate
- [ ] **Run the demo app** and try the changes[^6]

[^1]: _Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs._
[^2]: _While we aim for automated testing of the SDK, some aspects require manual testing. If you had to manually test 
something during development of this pull request, write those steps down._
[^3]: _While we are not looking for perfect coverage, the tool can point out potential cases that have been missed. Code coverage can be generated with: `./gradlew check` for Kotlin modules and `./gradlew connectedCheck -PIS_ANDROID_INSTRUMENTATION_TEST_COVERAGE_ENABLED=true` for Android modules._
[^4]: _Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit._
[^5]: _In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._
[^6]: _While the CI server runs the demo app to look for build failures or crashes, humans running the demo app are 
more likely to notice unexpected log messages, UI inconsistencies, or bad output data. Perform this step last, after verifying the code changes are safe to run locally._

### CODE REVIEW

**Review by LukasKorba — APPROVED** [review]

File: sdk-lib/src/main/java/cash/z/ecc/android/sdk/internal/TypesafeBackendImpl.kt [hunk 1]
```diff
@@ -114,21 +113,12 @@ internal class TypesafeBackendImpl(private val backend: Backend) : TypesafeBacke
...
             withContext(SdkDispatchers.DATABASE_IO) {
                 backend.getVerifiedTransparentBalance(address)
             }
-        val total =
-            withContext(SdkDispatchers.DATABASE_IO) {
-                backend.getTotalTransparentBalance(
-                    address
-                )
-            }
-        return WalletBalance(Zatoshi(total), Zatoshi(verified))
```
Comment (str4d) [inline sdk-lib/src/main/java/cash/z/ecc/android/sdk/internal/TypesafeBackendImpl.kt:131]: Other way round: for the change as implemented, we want to present the total balance within the address as available, so remove the call to `getVerifiedTransparentBalance` which uses the same confirmation depth of 10 as the shielded funds.

File: sdk-lib/src/main/java/cash/z/ecc/android/sdk/model/WalletBalance.kt [hunk 2]
```diff
@@ -1,28 +1,30 @@
...
-        require(total.value >= available.value) { "Wallet total balance must be >= available balance" }
-    }
+    /**
+     * The current total balance is calculated as a sum of [available], [changePending],
+     * and [valuePending].
+     */
+    val total = available + changePending + valuePending
 
+    /**
+     * The current pending balance is calculated as a distract of [total] and [available] balances.
```
Comment (str4d) [inline sdk-lib/src/main/java/cash/z/ecc/android/sdk/model/WalletBalance.kt:27]:
```suggestion
     * The current pending balance is calculated as the difference between [total] and [available] balances.
     * This is equivalent to the sum of [changePending] and [valuePending] balances.
```

**Review by str4d — CHANGES_REQUESTED** [review]
One blocking change, LGTM otherwise.

**Review by LukasKorba — APPROVED** [review]

File: sdk-lib/src/main/java/cash/z/ecc/android/sdk/internal/TypesafeBackendImpl.kt [hunk 3]
```diff
@@ -114,21 +113,12 @@ internal class TypesafeBackendImpl(private val backend: Backend) : TypesafeBacke
         backend.rewindBlockMetadataToHeight(height.value)
     }
 
-    override suspend fun getDownloadedUtxoBalance(address: String): WalletBalance {
-        // Note this implementation is not ideal because it requires two database queries without a transaction, which
-        // makes the data potentially inconsistent.  However the verified amount is queried first which makes this less
-        // bad.
+    override suspend fun getDownloadedUtxoBalance(address: String): Zatoshi {
         val verified =
```
Comment (str4d) [inline sdk-lib/src/main/java/cash/z/ecc/android/sdk/internal/TypesafeBackendImpl.kt:117]:
```suggestion
        val total =
```

File: sdk-lib/src/main/java/cash/z/ecc/android/sdk/internal/TypesafeBackendImpl.kt [hunk 4]
```diff
@@ -114,21 +113,12 @@ internal class TypesafeBackendImpl(private val backend: Backend) : TypesafeBacke
...
+                backend.getTotalTransparentBalance(address)
             }
-        val total =
-            withContext(SdkDispatchers.DATABASE_IO) {
-                backend.getTotalTransparentBalance(
-                    address
-                )
-            }
-        return WalletBalance(Zatoshi(total), Zatoshi(verified))
+        return Zatoshi(verified)
```
Comment (str4d) [inline sdk-lib/src/main/java/cash/z/ecc/android/sdk/internal/TypesafeBackendImpl.kt:121]:
```suggestion
        return Zatoshi(total)
```

**Review by str4d — APPROVED** [review]
utACK dce6773a77b02c350f0c78967a6962ca5de7b038 with fix.

**Pull request merged**